File: packages/sdk-java/src/main/java/com/ubag/sdk/UbagApiException.java
package com.ubag.sdk;

import java.util.Map;

/** Thrown when the gateway returns a non-2xx response. */
public class UbagApiException extends RuntimeException {

    private final int status;
    private final String method;
    private final String url;
    private final Map<String, String> headers;
    private final byte[] rawBody;
    private final Map<String, Object> envelope;

    public UbagApiException(
            int status,
            String method,
            String url,
            Map<String, String> headers,
            byte[] rawBody,
            Map<String, Object> envelope) {
        super(messageFor(status, envelope));
        this.status = status;
        this.method = method;
        this.url = url;
        this.headers = headers;
        this.rawBody = rawBody;
        this.envelope = envelope;
    }

    private static String messageFor(int status, Map<String, Object> envelope) {
        if (envelope != null) {
            Object error = envelope.get("error");
            if (error instanceof Map<?, ?> details) {
                Object message = details.get("message");
                if (message instanceof String text && !text.isEmpty()) {
                    return text;
                }
            }
        }
        return "UBAG API request failed with HTTP " + status;
    }

    public int status() {
        return status;
    }

    public String method() {
        return method;
    }

    public String url() {
        return url;
    }

    public Map<String, String> headers() {
        return headers;
    }

    public byte[] rawBody() {
        return rawBody;
    }

    public Map<String, Object> envelope() {
        return envelope;
    }

    @SuppressWarnings("unchecked")
    private Object errorField(String key) {
        if (envelope == null) {
            return null;
        }
        Object error = envelope.get("error");
        if (error instanceof Map<?, ?> details) {
            return ((Map<String, Object>) details).get(key);
        }
        return null;
    }

    public String code() {
        Object value = errorField("code");
        return value instanceof String text ? text : null;
    }

    public String category() {
        Object value = errorField("category");
        return value instanceof String text ? text : null;
    }

    public boolean retryable() {
        return Boolean.TRUE.equals(errorField("retryable"));
    }

    public String traceId() {
        Object value = errorField("trace_id");
        if (value instanceof String text && !text.isEmpty()) {
            return text;
        }
        String header = headers.get("ubag-trace-id");
        return header != null ? header : headers.get("x-request-id");
    }
}

File: packages/sdk-csharp/Ubag.Sdk/UbagApiException.cs
using System.Text.Json;

namespace Ubag.Sdk;

/// <summary>Thrown when the gateway returns a non-2xx response.</summary>
public sealed class UbagApiException : Exception
{
    public UbagApiException(
        int status,
        string httpMethod,
        string url,
        IReadOnlyDictionary<string, string> headers,
        string rawBody,
        JsonElement? envelope)
        : base(MessageFor(status, envelope))
    {
        Status = status;
        HttpMethod = httpMethod;
        Url = url;
        Headers = headers;
        RawBody = rawBody;
        Envelope = envelope;
    }

    public int Status { get; }

    public string HttpMethod { get; }

    public string Url { get; }

    public IReadOnlyDictionary<string, string> Headers { get; }

    public string RawBody { get; }

    public JsonElement? Envelope { get; }

    public string? Code => ErrorField("code");

    public string? Category => ErrorField("category");

    public bool Retryable =>
        Envelope is { } envelope
        && envelope.TryGetProperty("error", out var error)
        && error.TryGetProperty("retryable", out var retryable)
        && retryable.ValueKind == JsonValueKind.True;

    public string? TraceId
    {
        get
        {
            var fromEnvelope = ErrorField("trace_id");
            if (!string.IsNullOrEmpty(fromEnvelope))
            {
                return fromEnvelope;
            }

            if (Headers.TryGetValue("ubag-trace-id", out var traceHeader) && traceHeader.Length > 0)
            {
                return traceHeader;
            }

            return Headers.TryGetValue("x-request-id", out var requestId) && requestId.Length > 0
                ? requestId
                : null;
        }
    }

    private string? ErrorField(string name)
    {
        if (Envelope is { } envelope
            && envelope.TryGetProperty("error", out var error)
            && error.TryGetProperty(name, out var field)
            && field.ValueKind == JsonValueKind.String)
        {
            return field.GetString();
        }

        return null;
    }

    private static string MessageFor(int status, JsonElement? envelope)
    {
        if (envelope is { } element
            && element.TryGetProperty("error", out var error)
            && error.TryGetProperty("message", out var message)
            && message.ValueKind == JsonValueKind.String)
        {
            var text = message.GetString();
            if (!string.IsNullOrEmpty(text))
            {
                return text;
            }
        }

        return $"UBAG API request failed with HTTP {status}";
    }
}
